packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -0,0 +1,273 @@
+    if (widget.image.image != oldWidget.image.image)
+      _resolveImage();
+  }
+
+  @override
+  void reassemble() {
+    _resolveImage(); // in case the image cache was flushed
+    super.reassemble();
+  }
+
+  void _resolveImage() {
+    final ImageStream oldImageStream = _imageStream;
+    final Image image = widget.image;
+    _imageStream = image.image.resolve(createLocalImageConfiguration(
+        context,
+        size: image.width != null && image.height != null ? new Size(image.width, image.height) : null
+    ));
+    assert(_imageStream != null);
+
+    if (_imageStream.key != oldImageStream?.key) {
+      oldImageStream?.removeListener(_handleImageChanged);
+      if (!image.gaplessPlayback)
+        setState(() { _imageInfo = null; });
+      _imageStream.addListener(_handleImageChanged);
+    }
+
+    if (_phase == FadeInImagePhase.start)
+      _updatePhase();
+  }
+
+  void _handleImageChanged(ImageInfo imageInfo, bool synchronousCall) {
+    setState(() {
+      _imageInfo = imageInfo;
+      _updatePhase();
+    });
+  }
+
+  void _updatePhase() {
+    switch(_phase) {
+      case FadeInImagePhase.start:
+        if (_imageInfo != null)
+          _phase = FadeInImagePhase.completed;
+        else
+          _phase = FadeInImagePhase.waiting;
+        break;
+      case FadeInImagePhase.waiting:
+        if (_imageInfo != null) {
+          // Received image data. Begin placeholder fade-out.
+          _controller.duration = widget.fadeOutDuration;
+          _animation = new CurvedAnimation(
+            parent: _controller,
+            curve: widget.fadeOutCurve,
+          );
+          _phase = FadeInImagePhase.fadeOut;
+          _controller.reverse(from: 1.0);
+        }
+        break;
+      case FadeInImagePhase.fadeOut:
+        if (_controller.status == AnimationStatus.dismissed) {
+          // Done fading out placeholder. Begin target image fade-in.
+          _controller.duration = widget.fadeInDuration;
+          _animation = new CurvedAnimation(
+            parent: _controller,
+            curve: widget.fadeInCurve,
+          );
+          _phase = FadeInImagePhase.fadeIn;
+          _controller.forward(from: 0.0);
+        }
+        break;
+      case FadeInImagePhase.fadeIn:
+        if (_controller.status == AnimationStatus.completed) {
+          // Done finding in new image.
+          _phase = FadeInImagePhase.completed;
+        }
+        break;
+      case FadeInImagePhase.completed:
+      // Nothing to do.
+        break;
+    }
+  }
+
+  @override
+  void dispose() {
+    assert(_imageStream != null);
+    _imageStream.removeListener(_handleImageChanged);
+    _controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    assert(_phase != FadeInImagePhase.start);
+    if (_phase == FadeInImagePhase.waiting || _phase == FadeInImagePhase.fadeOut)
+      return _buildPlaceholderImage();
+    else
+      return _buildTargetImage();
```

Comment:
It's probably better to inline the methods. Factoring out parts of build methods is an anti-pattern because it hides the complexity of the build methods, and leads people away from optimising them (e.g. by factoring them into separate widgets).